Implement add the possibility to allow blueprints to add custom keywords.

mshima opened the pull request:
Fixes https://github.com/jhipster/jhipster-core/issues/448.

- Allows to externalise validation definitions.
- Allows to extend validation definitions.

Example for additional definitions:
```
importer.import(() => {
  return {
    applicationOptions: {
      clientFramework: {
        values: {
          ionic: 'ionic'
        }
      }
    }
  };
});
```

Please make sure the below checklist is followed for Pull Requests.
  - [ ] Checks are green
  - [ ] Tests are added where necessary
  - [ ] Documentation is added/updated where necessary
  - [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/jhipster-core/blob/master/CONTRIBUTING.md) are followed

mshima commented:
@MathieuAA let me know if I should keep working on this.

MathieuAA commented on lib/core/jhipster/index.js:
```diff
@@ -0,0 +1,96 @@
+
+class JHipsterOptions {
+  static get OPTIONS() {
+    return _.clone(builtInOptions);
+  }
+
+  constructor(blueprintOptions) {
+    if (blueprintOptions) {
+      this.options = _.merge({}, builtInOptions, blueprintOptions);
```
We could use destructuring instead of lodash, unless this requires some precise merging

mshima replied:
The merge should be precise.
```
{
  clientFramework: {
    angular: angular,
    react: react
  }
}
```
 +
```
{
  clientFramework: {
    vuejs: vuejs
  }
}
```
=

```
{
  clientFramework: {
    angular: angular,
    react: react,
    vuejs: vuejs
  }
}
```

MathieuAA commented on lib/core/jhipster/index.js:
```diff
@@ -0,0 +1,96 @@
+  /**
+   * Returns the option's type, one of string, boolean, list or integer.
+   * @param {String} optionName - the option's name.
+   * @returns {string} the option's type.
```
**S**tring

mshima replied:
All functions are copy from jdl_options.
Will fix them.

MathieuAA commented on lib/core/jhipster/index.js:
```diff
@@ -0,0 +1,96 @@
+  /**
+   * Checks whether the option value exists for the passed option name.
+   * @param {String} name - the option name.
+   * @param {String|Boolean|Number} value - the option value.
```
And array too, no?

MathieuAA commented on lib/validators/application_validator.js:
```diff
@@ -114,81 +107,90 @@ function checkForValidValues(jdlApplication) {
-function checkForUnknownApplicationOption(option) {
-  if (!doesOptionExist(option.name)) {
+function checkForUnknownApplicationOption(jdlApplication, option) {
+  if (!jdlApplication.jhipsterOptions.doesOptionExist(option.name)) {
```
`jdlApplication.jhipsterOptions.doesOptionExist` should be avoided: what if someone renames the jhipsterOptions attribute from the JDLApp? or removes it? We'd have to change this reference (and others) too.
Having a method in the JDLApplication is better (`jdlApplication.XXXX`). Maybe `jdlApplication.doesJHipsterOptionExist`.

mshima replied:
I will think about this one.
jhipsterOptions Is an instance of JHipsterOptions that contains all options definitions.
Unless I use JDLApplication directly instead of creating JHipsterOptions.

MathieuAA commented on lib/validators/application_validator.js:
```diff
@@ -114,81 +107,90 @@ function checkForValidValues(jdlApplication) {
-  if (getTypeForOption(option.name) === 'boolean' && typeof option.getValue() !== 'boolean') {
+function checkForBooleanValue(jdlApplication, option) {
+  if (
+    jdlApplication.jhipsterOptions.getTypeForOption(option.name) === 'boolean' &&
```
Same as above

MathieuAA commented:
@mshima I really like what you've done!!

mshima commented on lib/core/jhipster/application_options.js:
```diff
@@ -124,8 +124,7 @@ const optionValues = {
     angularX: 'angularX',
     angular: 'angular',
-    react: 'react',
-    vuejs: 'vuejs'
```
The test case adds vuejs option.
Once this is merged with a new generator-jhipster, a new vuejs blueprint version we will add the option to the blueprint.
Or just add again before merging.

mshima commented on lib/converters/parsedJDLToJDLObject/application_converter.js:
```diff
@@ -59,6 +59,9 @@ function addCustomValuesToApplication(parsedApplication, configuration) {
   if (configuration.creationTimestamp) {
     application.config.creationTimestamp = configuration.creationTimestamp;
   }
+  if (configuration.blueprintOptionsDefinition) {
```
It supports extending every built-in definition.

mshima commented on lib/core/jdl_application_factory.js:
```diff
@@ -38,23 +38,28 @@ module.exports = {
-        config: getConfigForMicroserviceApplication(baseConfig)
+        config: getConfigForMicroserviceApplication(baseConfig),
+        blueprintOptionsDefinition
       });
     case ApplicationTypes.GATEWAY:
       return new JDLApplication({
-        config: getConfigForGatewayApplication(baseConfig)
+        config: getConfigForGatewayApplication(baseConfig),
+        blueprintOptionsDefinition
       });
     case ApplicationTypes.UAA:
       return new JDLApplication({
-        config: getConfigForUAAApplication(baseConfig)
+        config: getConfigForUAAApplication(baseConfig),
+        blueprintOptionsDefinition
       });
     case ApplicationTypes.MONOLITH:
     default:
       return new JDLApplication({
-        config: getConfigForMonolithApplication(baseConfig)
+        config: getConfigForMonolithApplication(baseConfig),
+        blueprintOptionsDefinition
```
Fair enough.

mshima commented:
> Maybe we should have talked about what the `blueprintOptionsDefinition` should accept (its interface).
> Maybe some names are misleading or something. I don't know. What do you think?

We can discuss it later.
For now it accepts every definition.

MathieuAA commented:
I don't know, this could be considered as breaking change later...

mshima commented:
Currently can extend everything (methods should be extracted):
<details>
  <pre>
{
  applicationOptions: {
    ApplicationOptionTypes: {
      STRING: 'string',
      INTEGER: 'integer',
      BOOLEAN: 'boolean',
      LIST: 'list'
    },
    OptionTypes: {
      applicationType: [Object],
      authenticationType: [Object],
      baseName: [Object],
      blueprint: [Object],
      blueprints: [Object],
      buildTool: [Object],
      cacheProvider: [Object],
      clientFramework: [Object],
      clientPackageManager: [Object],
      clientTheme: [Object],
      clientThemeVariant: [Object],
      creationTimestamp: [Object],
      databaseType: [Object],
      devDatabaseType: [Object],
      dtoSuffix: [Object],
      embeddableLaunchScript: [Object],
      enableHibernateCache: [Object],
      enableSwaggerCodegen: [Object],
      enableTranslation: [Object],
      entitySuffix: [Object],
      experimental: [Object],
      i18N: [Object],
      installModules: [Object],
      jhiPrefix: [Object],
      jhipsterVersion: [Object],
      jwtSecretKey: [Object],
      languages: [Object],
      messageBroker: [Object],
      nativeLanguage: [Object],
      npm: [Object],
      otherModules: [Object],
      packageName: [Object],
      packageFolder: [Object],
      prodDatabaseType: [Object],
      reactive: [Object],
      rememberMeKey: [Object],
      searchEngine: [Object],
      serverPort: [Object],
      serviceDiscoveryType: [Object],
      skipClient: [Object],
      skipGit: [Object],
      skipInstall: [Object],
      skipServer: [Object],
      skipUserManagement: [Object],
      testFrameworks: [Object],
      uaaBaseName: [Object],
      useNpm: [Object],
      useSass: [Object],
      websocket: [Object],
      withEntities: [Object]
    },
    OptionNames: {
      APPLICATION_TYPE: 'applicationType',
      AUTHENTICATION_TYPE: 'authenticationType',
      BASE_NAME: 'baseName',
      BLUEPRINT: 'blueprint',
      BLUEPRINTS: 'blueprints',
      BUILD_TOOL: 'buildTool',
      CACHE_PROVIDER: 'cacheProvider',
      CLIENT_FRAMEWORK: 'clientFramework',
      CLIENT_PACKAGE_MANAGER: 'clientPackageManager',
      CLIENT_THEME: 'clientTheme',
      CLIENT_THEME_VARIANT: 'clientThemeVariant',
      CREATION_TIMESTAMP: 'creationTimestamp',
      DATABASE_TYPE: 'databaseType',
      DEV_DATABASE_TYPE: 'devDatabaseType',
      DTO_SUFFIX: 'dtoSuffix',
      EMBEDDABLE_LAUNCH_SCRIPT: 'embeddableLaunchScript',
      ENABLE_HIBERNATE_CACHE: 'enableHibernateCache',
      ENABLE_SWAGGER_CODEGEN: 'enableSwaggerCodegen',
      ENABLE_TRANSLATION: 'enableTranslation',
      ENTITY_SUFFIX: 'entitySuffix',
      EXPERIMENTAL: 'experimental',
      I_18_N: 'i18N',
      INSTALL_MODULES: 'installModules',
      JHI_PREFIX: 'jhiPrefix',
      JHIPSTER_VERSION: 'jhipsterVersion',
      JWT_SECRET_KEY: 'jwtSecretKey',
      LANGUAGES: 'languages',
      MESSAGE_BROKER: 'messageBroker',
      NATIVE_LANGUAGE: 'nativeLanguage',
      NPM: 'npm',
      OTHER_MODULES: 'otherModules',
      PACKAGE_NAME: 'packageName',
      PACKAGE_FOLDER: 'packageFolder',
      PROD_DATABASE_TYPE: 'prodDatabaseType',
      REACTIVE: 'reactive',
      REMEMBER_ME_KEY: 'rememberMeKey',
      SEARCH_ENGINE: 'searchEngine',
      SERVER_PORT: 'serverPort',
      SERVICE_DISCOVERY_TYPE: 'serviceDiscoveryType',
      SKIP_CLIENT: 'skipClient',
      SKIP_GIT: 'skipGit',
      SKIP_INSTALL: 'skipInstall',
      SKIP_SERVER: 'skipServer',
      SKIP_USER_MANAGEMENT: 'skipUserManagement',
      TEST_FRAMEWORKS: 'testFrameworks',
      UAA_BASE_NAME: 'uaaBaseName',
      USE_NPM: 'useNpm',
      USE_SASS: 'useSass',
      WEBSOCKET: 'websocket',
      WITH_ENTITIES: 'withEntities'
    },
    QuotedOptionNames: [ 'jhipsterVersion', 'rememberMeKey', 'jwtSecretKey' ],
    OptionValues: {
      applicationType: [Object],
      authenticationType: [Object],
      baseName: 'jhipster',
      blueprint: undefined,
      blueprints: [],
      buildTool: [Object],
      cacheProvider: [Object],
      clientFramework: [Object],
      clientPackageManager: [Object],
      clientTheme: 'none',
      clientThemeVariant: [Object],
      creationTimestamp: 1586121529477,
      databaseType: [Object],
      devDatabaseType: [Object],
      dtoSuffix: 'DTO',
      embeddableLaunchScript: true,
      enableHibernateCache: true,
      enableSwaggerCodegen: false,
      enableTranslation: true,
      entitySuffix: '',
      experimental: false,
      i18N: true,
      installModules: false,
      jhiPrefix: 'jhi',
      jhipsterVersion: '',
      jwtSecretKey: '',
      languages: [Array],
      messageBroker: [Object],
      nativeLanguage: 'en',
      npm: true,
      otherModules: [],
      packageFolder: 'com/mycompany/myapp',
      packageName: 'com.mycompany.myapp',
      prodDatabaseType: [Object],
      reactive: false,
      rememberMeKey: '',
      searchEngine: [Object],
      serverPort: '8080',
      serviceDiscoveryType: [Object],
      skipClient: false,
      skipGit: false,
      skipInstall: false,
      skipServer: false,
      skipUserManagement: false,
      testFrameworks: [Object],
      uaaBaseName: '../uaa',
      useNpm: true,
      useSass: false,
      websocket: [Object],
      withEntities: false
    }
  },
  applicationTypes: {
    MONOLITH: 'monolith',
    MICROSERVICE: 'microservice',
    UAA: 'uaa',
    GATEWAY: 'gateway',
    exists: [Function]
  },
  binaryOptions: {
    Options: {
      DTO: 'dto',
      SERVICE: 'service',
      PAGINATION: 'pagination',
      MICROSERVICE: 'microservice',
      SEARCH: 'search',
      ANGULAR_SUFFIX: 'angularSuffix',
      CLIENT_ROOT_FOLDER: 'clientRootFolder'
    },
    Values: {
      dto: [Object],
      service: [Object],
      pagination: [Object],
      search: [Object]
    },
    exists: [Function: exists],
    forEach: [Function: forEach]
  },
  databaseTypes: {
    SQL: 'sql',
    MYSQL: 'mysql',
    MARIADB: 'mariadb',
    POSTGRESQL: 'postgresql',
    MSSQL: 'mssql',
    ORACLE: 'oracle',
    MONGODB: 'mongodb',
    CASSANDRA: 'cassandra',
    COUCHBASE: 'couchbase',
    NEO4J: 'neo4j',
    NO: 'no',
    isSql: [Function]
  },
  defaultApplicationOptions: {
    getConfigForMonolithApplication: [Function: getConfigForMonolithApplication],
    getConfigForGatewayApplication: [Function: getConfigForGatewayApplication],
    getConfigForMicroserviceApplication: [Function: getConfigForMicroserviceApplication],
    getConfigForUAAApplication: [Function: getConfigForUAAApplication],
    getDefaultConfigForNewApplication: [Function: getDefaultConfigForNewApplication]
  },
  deploymentOptions: {
    Options: {
      deploymentType: [Object],
      gatewayType: [Object],
      monitoring: [Object],
      directoryPath: '../',
      appsFolders: [],
      clusteredDbApps: [],
      consoleOptions: [Object],
      serviceDiscoveryType: [Object],
      dockerRepositoryName: '',
      dockerPushCommand: 'docker push',
      kubernetesNamespace: 'default',
      kubernetesServiceType: [Object],
      ingressDomain: '',
      istio: [Object],
      openshiftNamespace: 'default',
      storageType: [Object],
      defaults: [Function]
    },
    DeploymentTypes: {
      DOCKERCOMPOSE: 'docker-compose',
      KUBERNETES: 'kubernetes',
      OPENSHIFT: 'openshift',
      exists: [Function: exists]
    }
  },
  fileTypes: {
    CommonDBTypes: {
      STRING: 'String',
      INTEGER: 'Integer',
      LONG: 'Long',
      BIG_DECIMAL: 'BigDecimal',
      FLOAT: 'Float',
      DOUBLE: 'Double',
      UUID: 'UUID',
      ENUM: 'Enum',
      BOOLEAN: 'Boolean',
      LOCAL_DATE: 'LocalDate',
      ZONED_DATE_TIME: 'ZonedDateTime',
      BLOB: 'Blob',
      ANY_BLOB: 'AnyBlob',
      IMAGE_BLOB: 'ImageBlob',
      TEXT_BLOB: 'TextBlob',
      INSTANT: 'Instant',
      DURATION: 'Duration'
    },
    CassandraTypes: {
      STRING: 'String',
      INTEGER: 'Integer',
      LONG: 'Long',
      BIG_DECIMAL: 'BigDecimal',
      FLOAT: 'Float',
      DOUBLE: 'Double',
      BOOLEAN: 'Boolean',
      DATE: 'Date',
      UUID: 'UUID',
      INSTANT: 'Instant',
      BYTE_BUFFER: 'ByteBuffer',
      ZONED_DATE_TIME: 'ZonedDateTime'
    },
    isCommonDBType: [Function: isCommonDBType],
    isCassandraType: [Function: isCassandraType],
    hasValidation: [Function: hasValidation],
    getIsType: [Function: getIsType]
  },
  relationshipOptions: {
    JPA_DERIVED_IDENTIFIER: 'jpaDerivedIdentifier',
    exists: [Function: exists],
    doesRelationshipOptionExist: [Function: exists]
  },
  relationshipTypes: {
    ONE_TO_ONE: 'OneToOne',
    ONE_TO_MANY: 'OneToMany',
    MANY_TO_ONE: 'ManyToOne',
    MANY_TO_MANY: 'ManyToMany',
    exists: [Function]
  },
  reservedKeywords: {
    isReserved: [Function: isReserved],
    isReservedClassName: [Function: isReservedClassName],
    isReservedTableName: [Function: isReservedTableName],
    isReservedFieldName: [Function: isReservedFieldName],
    JHIPSTER: [
      'ACCOUNT',      'ACTIVATE',
      'AUDITS',       'CONFIGURATION',
      'DOCS',         'HEALTH',
      'LOGS',         'METRICS',
      'PASSWORD',     'REGISTER',
      'RESET',        'SESSIONS',
      'SETTINGS',     'TEST',
      'EVENTMANAGER', 'PRINCIPAL',
      'ENTITY',       'RESULT'
    ],
    ANGULAR: [
      'CLASS',        'NODENAME',
      'NODETYPE',     'COMPONENT',
      'SUBSCRIPTION', 'RESPONSE',
      'OBSERVABLE',   'INJECTABLE',
      'HTTP',         'ROUTER'
    ],
    REACT: undefined,
    JAVA: [
      'ABSTRACT',     'CONTINUE',  'FOR',
      'NEW',          'SWITCH',    'ASSERT',
      'DEFAULT',      'GOTO',      'PACKAGE',
      'SYNCHRONIZED', 'BOOLEAN',   'DO',
      'IF',           'PRIVATE',   'THIS',
      'BREAK',        'DOUBLE',    'IMPLEMENTS',
      'PROTECTED',    'THROW',     'BYTE',
      'ELSE',         'IMPORT',    'PUBLIC',
      'THROWS',       'CASE',      'ENUM',
      'INSTANCEOF',   'RETURN',    'TRANSIENT',
      'CATCH',        'EXTENDS',   'INT',
      'SHORT',        'TRY',       'CHAR',
      'FINAL',        'INTERFACE', 'STATIC',
      'VOID',         'CLASS',     'FINALLY',
      'LONG',         'STRICTFP',  'VOLATILE',
      'CONST',        'FLOAT',     'NATIVE',
      'SUPER',        'WHILE'
    ],
    TYPESCRIPT: [
      'BREAK',     'CASE',       'CATCH',
      'CLASS',     'CONST',      'CONSTRUCTOR',
      'CONTINUE',  'DEBUGGER',   'DEFAULT',
      'DELETE',    'DO',         'ELSE',
      'ENUM',      'EXPORT',     'EXTENDS',
      'FALSE',     'FINALLY',    'FOR',
      'FUNCTION',  'IF',         'IMPORT',
      'IN',        'INSTANCEOF', 'NEW',
      'NULL',      'RETURN',     'SUPER',
      'SWITCH',    'THIS',       'THROW',
      'TRUE',      'TRY',        'TYPEOF',
      'VAR',       'VOID',       'WHILE',
      'WITH',      'IMPLEMENTS', 'INTERFACE',
      'LET',       'PACKAGE',    'PRIVATE',
      'PROTECTED', 'PUBLIC',     'STATIC',
      'YIELD'
    ],
    MYSQL: [
      'ACCESSIBLE',   'ADD',               'ALL',
      'ALTER',        'ANALYZE',           'AND',
      'AS',           'ASC',               'ASENSITIVE',
      'BEFORE',       'BETWEEN',           'BIGINT',
      'BINARY',       'BLOB',              'BOTH',
      'BY',           'CALL',              'CASCADE',
      'CASE',         'CHANGE',            'CHAR',
      'CHARACTER',    'CHECK',             'COLLATE',
      'COLUMN',       'CONDITION',         'CONSTRAINT',
      'CONTINUE',     'CONVERT',           'CREATE',
      'CROSS',        'CUME_DIST',         'CURRENT_DATE',
      'CURRENT_TIME', 'CURRENT_TIMESTAMP', 'CURRENT_USER',
      'CURSOR',       'DATABASE',          'DATABASES',
      'DAY_HOUR',     'DAY_MICROSECOND',   'DAY_MINUTE',
      'DAY_SECOND',   'DEC',               'DECIMAL',
      'DECLARE',      'DEFAULT',           'DELAYED',
      'DELETE',       'DENSE_RANK',        'DESC',
      'DESCRIBE',     'DETERMINISTIC',     'DISTINCT',
      'DISTINCTROW',  'DIV',               'DOUBLE',
      'DROP',         'DUAL',              'EACH',
      'ELSE',         'ELSEIF',            'EMPTY',
      'ENCLOSED',     'ESCAPED',           'EXCEPT',
      'EXISTS',       'EXIT',              'EXPLAIN',
      'FALSE',        'FETCH',             'FIRST_VALUE',
      'FLOAT',        'FLOAT4',            'FLOAT8',
      'FOR',          'FORCE',             'FOREIGN',
      'FROM',         'FULLTEXT',          'GRANT',
      'GROUP',        'GROUPING',          'GROUPS',
      'HAVING',       'HIGH_PRIORITY',     'HOUR_MICROSECOND',
      'HOUR_MINUTE',  'HOUR_SECOND',       'IF',
      'IGNORE',       'IN',                'INDEX',
      'INFILE',       'INNER',             'INOUT',
      'INSENSITIVE',  'INSERT',            'INT',
      'INT1',
      ... 156 more items
    ],
    POSTGRESQL: [
      'ALL',               'ANALYSE',         'ANALYZE',
      'AND',               'ANY',             'ARRAY',
      'AS',                'ASC',             'ASYMMETRIC',
      'AUTHORIZATION',     'BINARY',          'BOTH',
      'CASE',              'CAST',            'CHECK',
      'COLLATE',           'COLLATION',       'COLUMN',
      'CONCURRENTLY',      'CONSTRAINT',      'CREATE',
      'CROSS',             'CURRENT_CATALOG', 'CURRENT_DATE',
      'CURRENT_ROLE',      'CURRENT_SCHEMA',  'CURRENT_TIME',
      'CURRENT_TIMESTAMP', 'CURRENT_USER',    'DEFAULT',
      'DEFERRABLE',        'DESC',            'DISTINCT',
      'DO',                'ELSE',            'END',
      'EXCEPT',            'FALSE',           'FETCH',
      'FOR',               'FOREIGN',         'FROM',
      'FULL',              'GRANT',           'GROUP',
      'HAVING',            'ILIKE',           'IN',
      'INITIALLY',         'INNER',           'INTERSECT',
      'INTO',              'IS',              'ISNULL',
      'JOIN',              'LATERAL',         'LEADING',
      'LEFT',              'LIKE',            'LIMIT',
      'LOCALTIME',         'LOCALTIMESTAMP',  'NATURAL',
      'NOT',               'NOTNULL',         'NULL',
      'OFFSET',            'ON',              'ONLY',
      'OR',                'ORDER',           'OUTER',
      'OVERLAPS',          'PLACING',         'PRIMARY',
      'REFERENCES',        'RETURNING',       'RIGHT',
      'SELECT',            'SESSION_USER',    'SIMILAR',
      'SOME',              'SYMMETRIC',       'TABLE',
      'THEN',              'TO',              'TRAILING',
      'TRUE',              'UNION',           'UNIQUE',
      'USER',              'USING',           'VARIADIC',
      'VERBOSE',           'WHEN',            'WHERE',
      'WINDOW',            'WITH'
    ],
    CASSANDRA: [
      'ADD',         'ALL',          'ALTER',       'AND',
      'ANY',         'APPLY',        'AS',          'ASC',
      'ASCII',       'AUTHORIZE',    'BATCH',       'BEGIN',
      'BIGINT',      'BLOB',         'BOOLEAN',     'BY',
      'CLUSTERING',  'COLUMNFAMILY', 'COMPACT',     'CONSISTENCY',
      'COUNT',       'COUNTER',      'CREATE',      'DECIMAL',
      'DELETE',      'DESC',         'DOUBLE',      'DROP',
      'EACH_QUORUM', 'FLOAT',        'FROM',        'GRANT',
      'IN',          'INDEX',        'CUSTOM',      'INSERT',
      'INT',         'INTO',         'KEY',         'KEYSPACE',
      'LEVEL',       'LIMIT',        'LOCAL_ONE',   'LOCAL_QUORUM',
      'MODIFY',      'NORECURSIVE',  'NOSUPERUSER', 'OF',
      'ON',          'ONE',          'ORDER',       'PASSWORD',
      'PERMISSION',  'PERMISSIONS',  'PRIMARY',     'QUORUM',
      'REVOKE',      'SCHEMA',       'SELECT',      'SET',
      'STORAGE',     'SUPERUSER',    'TABLE',       'TEXT',
      'TIMESTAMP',   'TIMEUUID',     'THREE',       'TOKEN',
      'TRUNCATE',    'TTL',          'TWO',         'TYPE',
      'UPDATE',      'USE',          'USER',        'USERS',
      'USING',       'UUID',         'VALUES',      'VARCHAR',
      'VARINT',      'WHERE',        'WITH',        'WRITETIME',
      'DISTINCT',    'BYTE',         'SMALLINT',    'COMPLEX',
      'ENUM',        'DATE',         'INTERVAL',    'MACADDR',
      'BITSTRING'
    ],
    COUCHBASE: [
      'ALL',        'AND',        'ANY',        'APPLY',
      'AS',         'ASC',        'AT',         'AUTOGENERATED',
      'BETWEEN',    'BTREE',      'BUCKET',     'BY',
      'CASE',       'CLOSED',     'CREATE',     'COMPACTION',
      'COMPACT',    'CONNECT',    'CONNECTED',  'CORRELATE',
      'COLLECTION', 'DATASET',    'DATAVERSE',  'DECLARE',
      'DEFINITION', 'DECLARE',    'DEFINITION', 'DELETE',
      'DESC',       'DISCONNECT', 'DISTINCT',   'DROP',
      'ELEMENT',    'EXPLAIN',    'ELSE',       'ENFORCED',
      'END',        'EVERY',      'EXCEPT',     'EXISTS',
      'EXTERNAL',   'FEED',       'FILTER',     'FLATTEN',
      'FOR',        'FROM',       'FULL',       'FUNCTION',
      'GROUP',      'HAVING',     'HINTS',      'IF',
      'INTO',       'IN',         'INDEX',      'INGESTION',
      'INNER',      'INSERT',     'INTERNAL',   'INTERSECT',
      'IS',         'JOIN',       'KEYWORD',    'KEY',
      'LEFT',       'LETTING',    'LET',        'LIKE',
      'LIMIT',      'LOAD',       'NODEGROUP',  'NGRAM',
      'NOT',        'OFFSET',     'ON',         'OPEN',
      'OR',         'ORDER',      'OUTER',      'OUTPUT',
      'PATH',       'POLICY',     'PRE-SORTED', 'PRIMARY',
      'RAW',        'REFRESH',    'RETURN',     'RTREE',
      'RUN',        'SATISFIES',  'SECONDARY',  'SELECT',
      'SET',        'SHADOW',     'SOME',       'TEMPORARY',
      'THEN',       'TYPE',       'TO',         'UNION',
      ... 10 more items
    ],
    ORACLE: [
      'ACCESS',
      'ACCOUNT',
      'ACTIVATE',
      'ADD',
      'ADMIN',
      'ADVISE',
      'AFTER',
      'ALL',
      'ALL_ROWS',
      'ALLOCATE',
      'ALTER',
      'ANALYZE',
      'AND',
      'ANY',
      'ARCHIVE',
      'ARCHIVELOG',
      'ARRAY',
      'AS',
      'ASC',
      'AT',
      'AUDIT',
      'AUTHENTICATED',
      'AUTHORIZATION',
      'AUTOEXTEND',
      'AUTOMATIC',
      'BACKUP',
      'BECOME',
      'BEFORE',
      'BEGIN',
      'BETWEEN',
      'BFILE',
      'BITMAP',
      'BLOB',
      'BLOCK',
      'BODY',
      'BY',
      'CACHE',
      'CACHE_INSTANCES',
      'CANCEL',
      'CASCADE',
      'CAST',
      'CFILE',
      'CHAINED',
      'CHANGE',
      'CHAR',
      'CHAR_CS',
      'CHARACTER',
      'CHECK',
      'CHECKPOINT',
      'CHOOSE',
      'CHUNK',
      'CLEAR',
      'CLOB',
      'CLONE',
      'CLOSE',
      'CLOSE_CACHED_OPEN_CURSORS',
      'CLUSTER',
      'COALESCE',
      'COLUMN',
      'COLUMNS',
      'COMMENT',
      'COMMIT',
      'COMMITTED',
      'COMPATIBILITY',
      'COMPILE',
      'COMPLETE',
      'COMPOSITE_LIMIT',
      'COMPRESS',
      'COMPUTE',
      'CONNECT',
      'CONNECT_TIME',
      'CONSTRAINT',
      'CONSTRAINTS',
      'CONTENTS',
      'CONTINUE',
      'CONTROLFILE',
      'CONVERT',
      'COST',
      'CPU_PER_CALL',
      'CPU_PER_SESSION',
      'CREATE',
      'CURRENT',
      'CURRENT_SCHEMA',
      'CURREN_USER',
      'CURSOR',
      'CYCLE',
      ' ',
      'DANGLING',
      'DATABASE',
      'DATAFILE',
      'DATAFILES',
      'DATAOBJNO',
      'DATE',
      'DBA',
      'DBHIGH',
      'DBLOW',
      'DBMAC',
      'DEALLOCATE',
      'DEBUG',
      'DEC',
      ... 381 more items
    ],
    MONGODB: [ 'DOCUMENT' ],
    MSSQL: [
      'ADD',               'ALL',             'ALTER',
      'AND',               'ANY',             'AS',
      'ASC',               'AUTHORIZATION',   'BACKUP',
      'BEGIN',             'BETWEEN',         'BREAK',
      'BROWSE',            'BULK',            'BY',
      'CASCADE',           'CASE',            'CHECK',
      'CHECKPOINT',        'CLOSE',           'CLUSTERED',
      'COALESCE',          'COLLATE',         'COLUMN',
      'COMMIT',            'COMPUTE',         'CONSTRAINT',
      'CONTAINS',          'CONTAINSTABLE',   'CONTINUE',
      'CONVERT',           'CREATE',          'CROSS',
      'CURRENT',           'CURRENT_DATE',    'CURRENT_TIME',
      'CURRENT_TIMESTAMP', 'CURRENT_USER',    'CURSOR',
      'DATABASE',          'DBCC',            'DEALLOCATE',
      'DECLARE',           'DEFAULT',         'DELETE',
      'DENY',              'DESC',            'DISK',
      'DISTINCT',          'DISTRIBUTED',     'DOUBLE',
      'DROP',              'DUMP',            'ELSE',
      'END',               'ERRLVL',          'ESCAPE',
      'EXCEPT',            'EXEC',            'EXECUTE',
      'EXISTS',            'EXIT',            'EXTERNAL',
      'FETCH',             'FILE',            'FILLFACTOR',
      'FOR',               'FOREIGN',         'FREETEXT',
      'FREETEXTTABLE',     'FROM',            'FULL',
      'FUNCTION',          'GOTO',            'GRANT',
      'GROUP',             'HAVING',          'HOLDLOCK',
      'IDENTITY',          'IDENTITY_INSERT', 'IDENTITYCOL',
      'IF',                'IN',              'INDEX',
      'INNER',             'INSERT',          'INTERSECT',
      'INTO',              'IS',              'JOIN',
      'KEY',               'KILL',            'LEFT',
      'LIKE',              'LINENO',          'LOAD',
      'MERGE',             'NATIONAL',        'NOCHECK',
      'NONCLUSTERED',
      ... 85 more items
    ],
    NEO4J: [
      'CALL',    'CREATE', 'DELETE',     'DETACH',     'EXISTS',
      'FOREACH', 'LOAD',   'MATCH',      'MERGE',      'OPTIONAL',
      'REMOVE',  'RETURN', 'SET',        'START',      'UNION',
      'UNWIND',  'WITH',   'LIMIT',      'ORDER',      'SKIP',
      'WHERE',   'YIELD',  'ASC',        'ASCENDING',  'ASSERT',
      'BY',      'CSV',    'DESC',       'DESCENDING', 'ON',
      'ALL',     'CASE',   'ELSE',       'END',        'THEN',
      'WHEN',    'AND',    'AS',         'CONTAINS',   'DISTINCT',
      'ENDS',    'IN',     'IS',         'NOT',        'OR',
      'STARTS',  'XOR',    'CONSTRAINT', 'CREATE',     'DROP',
      'EXISTS',  'INDEX',  'NODE',       'KEY',        'UNIQUE',
      'INDEX',   'JOIN',   'PERIODIC',   'COMMIT',     'SCAN',
      'USING',   'false',  'null',       'true',       'ADD',
      'DO',      'FOR',    'MANDATORY',  'OF',         'REQUIRE',
      'SCALAR'
    ]
  },
  unaryOptions: {
    SKIP_CLIENT: 'skipClient',
    SKIP_SERVER: 'skipServer',
    NO_FLUENT_METHOD: 'noFluentMethod',
    READ_ONLY: 'readOnly',
    FILTER: 'filter',
    EMBEDDED: 'embedded',
    forEach: [Function],
    exists: [Function]
  },
  validations: {
    REQUIRED: 'required',
    UNIQUE: 'unique',
    MIN: 'min',
    MAX: 'max',
    MINLENGTH: 'minlength',
    MAXLENGTH: 'maxlength',
    PATTERN: 'pattern',
    MINBYTES: 'minbytes',
    MAXBYTES: 'maxbytes',
    exists: [Function],
    needsValue: [Function]
  }
}
  </pre>
  </details>

mshima commented:
@MathieuAA I think is ready, need reviews, I will look again after some break.
Workflow needed by https://github.com/jhipster/jhipster-core/issues/415 is also implemented:
1 - generator-jhipster executes `makeJDLImporter()` (JDL creating moved to this function, validation is deferred)
2 - generator-jhipster gets blueprints from the jdl.
3 - generator-jhipster lookup for the blueprints.
4 - generator-jhipster gets the jdl definition from blueprints.
5 - generator-jhipster register the definition to the JDLApplication `extendDefinitions()`
6 - generator-jhipster executes `import()` (validation)

Once this is ready to merge we should discuss the jdl definition format for the blueprints.

One kind of resolved problem is a conflict between lexer custom options and step 1.
Why?
`Cannot get blueprint jdl definition without blueprint names`, `cannot get blueprint names without a parsed jdl`  and `cannot get a parsed jdl without blueprint jdl definition`.
The only solution I can think of is making jdl options unique by some way.
I implemented that options like `foo:bar` with `:` should be ignored at step 1 (lexer), and only validated at step 6.

MathieuAA commented:
> 1 - generator-jhipster executes makeJDLImporter() (JDL creating moved to this function, validation is deferred)
2 - generator-jhipster gets blueprints from the jdl.
3 - generator-jhipster lookup for the blueprints.
4 - generator-jhipster gets the jdl definition from blueprints.
5 - generator-jhipster register the definition to the JDLApplication extendDefinitions()
6 - generator-jhipster executes import() (validation)

That means breaking everything, unless I'm missing the point here.
I imagined something more like:
  1. importer creation with JDL content containing the `blueprints` attributes in applications
    - nothing happens here, the importer is just created and ready to answer to `importer` calls
  1. `import` is called, and the import state is returned
    - the import state possesses a list of exported apps, entities & deployments, but also a list of blueprints for each application
    - like
```json
{
  "blueprintsForEachApplication": {
    "app1": ["generator-jhipster-vuejs", "generator-jhipster-dotnetcore"]
  }
}
```
  1. the generator looks up for the blueprints
    - fails miserably if the user didn't install them or couldn't find them for some other reasons (incidentally, this should be documented)
    - it picks up the versions, installs the blueprints for each app, which modifies the relevant files for each app
  1. done

mshima commented:
> > 1 - generator-jhipster executes makeJDLImporter() (JDL creating moved to this function, validation is deferred)
> > 2 - generator-jhipster gets blueprints from the jdl.
> > 3 - generator-jhipster lookup for the blueprints.
> > 4 - generator-jhipster gets the jdl definition from blueprints.
> > 5 - generator-jhipster register the definition to the JDLApplication extendDefinitions()
> > 6 - generator-jhipster executes import() (validation)
> 
> That means breaking everything, unless I'm missing the point here.
> I imagined something more like:
> 
> 1. importer creation with JDL content containing the `blueprints` attributes in applications
>    - nothing happens here, the importer is just created and ready to answer to `importer` calls
> 2. `import` is called, and the import state is returned
>    - the import state possesses a list of exported apps, entities & deployments, but also a list of blueprints for each application
>    - like
> 
> ```json
> {
>   "blueprintsForEachApplication": {
>     "app1": ["generator-jhipster-vuejs", "generator-jhipster-dotnetcore"]
>   }
> }
> ```
> 
> 1. the generator looks up for the blueprints
>    - fails miserably if the user didn't install them or couldn't find them for some other reasons (incidentally, this should be documented)
>    - it picks up the versions, installs the blueprints for each app, which modifies the relevant files for each app
> 2. done

Don’t validate blueprint configuration at all?
How to differ blueprint configuration from normal ones?
What should we validate?

MathieuAA commented:
The more I think about it, the more I think that validating complicates everything, without real added value. Maybe the validations shouldn't be implemented in JCore.

mshima commented:
Removing validation should make much more easier.
Validation isn’t required for jdl studio?

mshima commented:
Removing validation makes this PR useless.
This decision should be taken ASAP.

MathieuAA commented:
The studio needs validation but it can always be decoupled from the rest. I'm in. If you're okay with that too, let's talk about how we want to proceed on all this to try preventing this kind of thing happening again and be on the right track from now on.

mshima commented:
This PR allows to move rules to generator-jhipster.
I think if we continue this approach future changes to jhipster-core will be rare.

mshima commented:
@MathieuAA it's done.
The missing part is to allow blueprint configs without `:` or define a better pattern.

MathieuAA commented:
I struggling to understand everything here, can you give me a big picture of this solution?
For instance, I don't get how making the configuration options generic fits in the mix.

mshima commented:
> I struggling to understand everything here, can you give me a big picture of this solution?
> For instance, I don't get how making the configuration options generic fits in the mix.

I think you should review https://github.com/jhipster/jhipster-core/pull/453 first.
It’s the first commit.

It’s needed to export.
If we don’t create a generic field here, then the value is lost and will not be exported.

The others changes is to centralize validation and create a simple structure for externalize and simplify validations.

mshima commented:
Definitions format that can be moved to generator-jhipster
<details>
  <pre>
{
  applicationOptions: {
    applicationType: { label: 'applicationType', type: 'string', values: [Object] },
    authenticationType: { label: 'authenticationType', type: 'string', values: [Object] },
    baseName: { label: 'baseName', type: 'string', values: 'jhipster' },
    blueprint: { label: 'blueprint', type: 'string', values: undefined },
    blueprints: { label: 'blueprints', type: 'list', values: [], alwaysOmit: true },
    buildTool: { label: 'buildTool', type: 'string', values: [Object] },
    cacheProvider: { label: 'cacheProvider', type: 'string', values: [Object] },
    clientFramework: { label: 'clientFramework', type: 'string', values: [Object] },
    clientPackageManager: { label: 'clientPackageManager', type: 'string', values: [Object] },
    clientTheme: { label: 'clientTheme', type: 'string', values: 'none' },
    clientThemeVariant: {
      label: 'clientThemeVariant',
      type: 'string',
      values: [Object],
      omitOnNegativeValue: true
    },
    creationTimestamp: {
      label: 'creationTimestamp',
      type: 'integer',
      values: 1586615685149
    },
    databaseType: { label: 'databaseType', type: 'string', values: [Object] },
    devDatabaseType: { label: 'devDatabaseType', type: 'string', values: [Object] },
    dtoSuffix: {
      label: 'dtoSuffix',
      type: 'string',
      values: 'DTO',
      omitOnNegativeValue: true
    },
    embeddableLaunchScript: { label: 'embeddableLaunchScript', type: 'boolean', values: true },
    enableHibernateCache: { label: 'enableHibernateCache', type: 'boolean', values: true },
    enableSwaggerCodegen: { label: 'enableSwaggerCodegen', type: 'boolean', values: false },
    enableTranslation: { label: 'enableTranslation', type: 'boolean', values: true },
    entitySuffix: {
      label: 'entitySuffix',
      type: 'string',
      values: '',
      omitOnNegativeValue: true
    },
    experimental: { label: 'experimental', type: 'boolean', values: false },
    i18N: { label: 'i18N', type: 'boolean', values: true },
    installModules: { label: 'installModules', type: 'boolean', values: false },
    jhiPrefix: { label: 'jhiPrefix', type: 'string', values: 'jhi' },
    jhipsterVersion: {
      label: 'jhipsterVersion',
      type: 'string',
      values: '',
      quoted: true
    },
    jwtSecretKey: { label: 'jwtSecretKey', type: 'string', values: '', quoted: true },
    languages: { label: 'languages', type: 'list', values: [Array] },
    messageBroker: { label: 'messageBroker', type: 'string', values: [Object] },
    nativeLanguage: { label: 'nativeLanguage', type: 'string', values: 'en' },
    npm: { label: 'npm', type: 'boolean', values: true },
    otherModules: { label: 'otherModules', type: 'list', values: [] },
    packageName: {
      label: 'packageName',
      type: 'string',
      values: 'com.mycompany.myapp'
    },
    packageFolder: {
      label: 'packageFolder',
      type: 'string',
      values: 'com/mycompany/myapp',
      alwaysOmit: true
    },
    prodDatabaseType: { label: 'prodDatabaseType', type: 'string', values: [Object] },
    reactive: { label: 'reactive', type: 'boolean', values: false },
    rememberMeKey: {
      label: 'rememberMeKey',
      type: 'string',
      values: '',
      quoted: true
    },
    searchEngine: { label: 'searchEngine', type: 'string', values: [Object] },
    serverPort: { label: 'serverPort', type: 'integer', values: '8080' },
    serviceDiscoveryType: { label: 'serviceDiscoveryType', type: 'string', values: [Object] },
    skipClient: { label: 'skipClient', type: 'boolean', values: false },
    skipGit: { label: 'skipGit', type: 'boolean', values: false },
    skipInstall: { label: 'skipInstall', type: 'boolean', values: false },
    skipServer: { label: 'skipServer', type: 'boolean', values: false },
    skipUserManagement: { label: 'skipUserManagement', type: 'boolean', values: false },
    testFrameworks: { label: 'testFrameworks', type: 'list', values: [Object] },
    uaaBaseName: { label: 'uaaBaseName', type: 'string', values: '../uaa' },
    useNpm: { label: 'useNpm', type: 'boolean', values: true },
    useSass: { label: 'useSass', type: 'boolean', values: false },
    websocket: { label: 'websocket', type: 'string', values: [Object] },
    withEntities: { label: 'withEntities', type: 'boolean', values: false }
  },
  databases: {
    no: {},
    sql: { sql: true },
    mysql: { sql: true, prod: true },
    postgresql: { sql: true, prod: true },
    mariadb: { sql: true, prod: true },
    oracle: { sql: true, prod: true },
    mssql: { sql: true, prod: true },
    mongodb: { cannotMixProdDev: true },
    couchbase: { cannotMixProdDev: true },
    cassandra: { cannotMixProdDev: true },
    neo4j: { cannotMixProdDev: true },
    h2Memory: { sql: true, devOnly: true },
    h2Disk: { sql: true, devOnly: true }
  }
}
</pre>
</details>

MathieuAA commented:
@mshima a small warning, there may not be a need to do that just yet. If JCore and the generator are gonna be merged, this may not be the best solution. I don't know what the solution might be, but there may be changes. Let's wait a bit for that

mshima commented:
> @mshima a small warning, there may not be a need to do that just yet. If JCore and the generator are gonna be merged, this may not be the best solution. I don't know what the solution might be, but there may be changes. Let's wait a bit for that

In my opinion this is a must have to merge both.
It's needed to unify.

mshima commented:
Ok then. I will stop working on this.

mshima commented:
@MathieuAA do you want me to do something with this PR?
Remove some functionality?
Or should I close it?

MathieuAA commented:
We can close it I think

MathieuAA commented:
Yeah we can close it
